---
title: "Using mixed-effect model to investigate for the maximum observed burst size in Plasmoidum"
author: "Damie Pak"
date: "`r Sys.Date()`"
output: pdf_document
editor_options:
  markdown:
    wrap: 72
  chunk_output_type: console
bibliography: references.bib
---

```{r include=FALSE}
set.seed(24601)
library(here)
library(ggplot2)
library(brms)
library(performance)
library(bayesplot)
library(purrr)
library(dplyr)
library(tidybayes)
source(here("Code","Functions","FUNC_Package_Loader.R"))
source(here("Code", "Functions","FUNC_renamer_label.R"))
source(here("Code", "Functions", "FUNC_util_remove_underscore.R"))
load(here("Output", "brms_output", "fit_host_only.R"))
load(here("Output", "brms_output", "fit_par_only.R"))
load(here("Output", "brms_output", "fit_orig.R"))
source(here("Code", "Data_Analysis","00_full_phylogeny_tree.R"))
```

## The host phylogentic trees

Using the provided taxonomic subsetting tool, we obtain published
phylogeny trees for the host species of interest from VertLife.org
[@tonini2016, @upham2019, @jetz2012]. All phylogenetic trees are
provided as .nex files:

```{r include=FALSE}
###Reptile (Squamatian)
REP_Phylo <- read.nexus(here('Data',"Reptile","reptile_phylo_1000.nex")) 
###AVIAN (Birds)
AVE_Phylo <- read.nexus(here("Data", "Avian", "avian_phylo_1000.nex"))

###MAMMAL
MAM_Phylo <- read.nexus(here("Data", "Mammal", "mammal_phylo_1000.nex"))
```

Because the phylogenetic trees are constructed using Bayesian methods,
each Nexus file contains a distribution of possible trees ($n$ = 1000).
We construct a main consensus tree for each host clade using the 50%
majority rule and randomly resolving polytomies. Note that there are no
branch lengths when creating the consensus tree.

```{r}
###(50% majority rule consensus trees)
consensus_AVE_Tree <- consensus(AVE_Phylo, p = 0.5, check.labels = TRUE, rooted = TRUE)
consensus_REP_Tree <- consensus(REP_Phylo, p = 0.5, check.labels = TRUE, rooted = TRUE)
consensus_MAM_Tree <- consensus(MAM_Phylo, p = 0.5, check.labels = TRUE, rooted = TRUE)

###howManyPolytomies(consensus_MAM_Tree) #1 poly
###howManyPolytomies(consensus_REP_Tree) #2 poly
###howManyPolytomies(consensus_AVE_Tree) #None

###Resolving polytomies
consensus_AVE_Tree <- multi2di(consensus_AVE_Tree, random = TRUE)
consensus_REP_Tree <- multi2di(consensus_REP_Tree, random = TRUE)
consensus_MAM_Tree <- multi2di(consensus_MAM_Tree, random = TRUE)
```

To analyze all \textit{Plasmodium} species, we need to create a
supertree of the avian, squamatian, and mammalian hosts. First, we
create a template supertree:

```{r}
tip.labels <- c("mam", "birds", "squam")

edge <- matrix(c(4,1,
                 4,5,
                 5,2,
                 5,3), 
               byrow=TRUE, ncol=2)

edge.length<- c(1,1,1,1)

Nnode <-2

ordertree <- list(edge=edge, Nnode=Nnode, tip.label=tip.labels )
class(ordertree) <- 'phylo'
```

```{r}
plot(ordertree)
```

We then graft our three phylogenetic trees to this supertree:

```{r}

tree_list <- list(squam=consensus_REP_Tree, 
                  birds=consensus_AVE_Tree, 
                  mam= consensus_MAM_Tree)

class(tree_list) <- "multiPhylo"

###Graft on the reptile first.
Order_1_Reptile <- bind.tree(x = ordertree, 
                             y = consensus_REP_Tree, 
                             where = 3, interactive = FALSE)

###Graft on the birds second 
Order_2_Reptile_Avian  <- bind.tree(x = Order_1_Reptile, 
                                    y = consensus_AVE_Tree, 
                                    where = 2, 
                                    interactive = FALSE)

###Graft the mammals third.
Full_SuperTree_Host <- bind.tree(x = Order_2_Reptile_Avian , 
                                 y = consensus_MAM_Tree, 
                                 where = 1, 
                                 interactive = FALSE)
```

## The parasite phylogenetic tree

For the \textit{Plasmodium} phylogeny, we use the single phylogeny tree
provided by Galen et al. 2018 [@galen2018]:

```{r}
plasmodium_tree <- read.nexus(here(file = "Data", "16_amino_acid_partitioned_BEAST_relaxed_clock.tre"))

### Non plasmoidum species
### 1-20, 37-38,54-59 
###Example print(plasmodium_tree$tip.label)
### These are non-Plasmodium species
Non_Plas_species <- plasmodium_tree$tip.label[c(1:20, 28:29, 37:39, 54:59)] 
### These are the Plasmodium species
Plas_species <- plasmodium_tree$tip.label[-c(1:20, 28:29, 37:39, 54:59)]

### This is the tree with only Plasmodium species
plasmodium_tree_full <- drop.tip(plasmodium_tree, Non_Plas_species)

### Giving the tree a new label- CHECK to make sure!
new_Plasmoidum_Species_Label <- c(
  "azurophilum","chiricahuae","coatneyi", 
  "cyclopsi", "floridense", "fragile", 
  "gaboni", "lacertiliae", "leucocytica", 
  "mackerrasae", "malariae", "mexicanum", 
  "minuoviridae", "ovale", "berghei", 
  "chabaudi", "cynomolgi", "falciparum", 
  "gallinaceum", "giganteum", "inui", 
  "juxtanucleare", "knowlesi", "reichenowi", 
  "relictum", "vinckei", "vivax", "yoelii"
)

plasmodium_tree_full <- rename_labels(tree = plasmodium_tree_full, vec = new_Plasmoidum_Species_Label)

###DROP THESE THREE SPECIES
plasmodium_tree_full <- drop.tip(plasmodium_tree_full, c(
  "minuoviridae",
  "leucocytica",
  "gaboni"
))
```

We prune the tree to only contain the \textit{Plasmodium species} and
exclude \textit{P. minuoviridae}, \textit{P. leuocytica}. and
\textit{P.gaboni} due to the lack of morphological data.

## Investigating the phylogenetic signal

### Full mathematical formulation

An aim of this paper is to investigate if the maximum observed burst
size is explained by the parasite and/or host phylogeny. We construct a
generalized mixed-effect regression model with the package {brms}.
Because the maximum observed burst size is a positive integer, we assume
a poisson distribution with the default log-link function. The
'clustering' of observations due to the shared evolutionary
relationships is explained as random effects. The data that we use has
the \textit{Plasmodium} species and their observed burst size across
different hosts (experimental and natural).

```{r include=FALSE}
mal_dat_asex <- read.csv(here("Data","MALARIA_PHYLOGENY_HOST_VARIOUS.csv"))
#add the underscore between genus and species name 
mal_dat_asex$Host <- add_underscore(mal_dat_asex$Host) 
```

Example: For Plasmodium floridense, the species can differ in the
maximum observedburst sizes depending on the host!

```{r}
print(mal_dat_asex[20:30,c("Plasmodium_species","Host", "Upper_Burst_Size")])
```

To test for the phylogenetic signal, we create a random intercept model.
Here, we describe the observation ($i$) belonging to the
\textit{Plasmodium} group $j$ and host group $k$ to describe the maximum
observed burst size as $(Y_{j{i}k{i}})$. The full model has five random
effects: the parasite ($\alpha$) and host ($\phi$) membership accounting
for phylogeny, parasite ($\gamma$) and host ($\delta$) species name
accounting for the 'ecology' of the species (unaccounted by the
phylogenetic relatedness), and the individual observations ($\eta$). By
including the individualobservations as a random-effect, we can account
for the dispersion in the data [@harrison2014]. The full model
formulations are below:

$$
\begin{aligned}
  \operatorname{Y}_{i}  &\sim \operatorname{Poisson}(\lambda_i) \\
    \log(\lambda_i) &= \beta_0 + \alpha_{j[i]} + \phi_{k[i]} + \gamma_{k[i]} + \delta_{j[i]} + \eta_i \\
    \alpha_{j[i]}  &\sim N \left(0, \sigma_{\alpha}^2\Sigma_{\alpha} \right)
    \text{, for Plasmodium species (phy) j = 1,} \dots \text{, 24} \\
    \phi_{k[i]}  &\sim N \left(0, \sigma_{\phi}^2\Sigma_{\phi} \right)
    \text{, for Host species (phy) k = 1,} \dots \text{,52}\\
  \gamma_{j[i]}  &\sim N \left(0, \sigma_\gamma^2 \right)
    \text{, for Plasmodium species (eco) j = 1,} \dots \text{,24}\\
 \delta_{k[i]}  &\sim N \left(0, \sigma_\delta^2 \right)
    \text{, for Host species (eco) j = 1,} \dots \text{,52}\\
 \eta_{i}  &\sim N \left(0, \sigma_\eta^2 \right)
    \text{, for Observation  j = 1,} \dots \text{,79}\\
\end{aligned}
$$

where $B_0$ is the grand intercept, $\alpha_{j[i]}$ and $\phi_k[i]$ are
the random effects of the parasite and host phylogenies with
$\Sigma_\alpha$ and $\Sigma_\phi$ describing the variance-covariance
matrices calculated from phylogenetic trees (describes the correlation
structure between observations $i$). For $\gamma_{j[i]}$ and
$\delta_{k[i]}$, these are the random effects due to the species name
with variance $\sigma_{\gamma}^2$ and $\sigma_{\delta}^2$, respectively.
The individual-level random effect is then explained by the variance
$\sigma_\eta^2$.

## Running the full model in brms

### Calculating the covariance matrices

We use the compute.brlen to calculate the branch length and the vcv to
construct the variance covariance matrices.

```{r include=FALSE}
type_host_tree_subsetted <- drop.tip(Full_SuperTree_Host, 
                                     Full_SuperTree_Host$tip.label
                                     [-match(mal_dat_asex$Host, 
                                             Full_SuperTree_Host$tip.label)])
```

```{r}
vcv_Plasmodium <- vcv(compute.brlen(plasmodium_tree_full))
vcv_TypeHost <- vcv(compute.brlen(type_host_tree_subsetted))
```

Having more species on the phylogenetic tree increases the overall
branch lengths. For comparison between the host and parasite effects, we
standardize the variance covariance matrices by dividing each matrix by
its determinant:

```{r}
vcv_Plasmodium_standard <- vcv_Plasmodium /(det(vcv_Plasmodium )^(1/nrow(vcv_Plasmodium)))

vcv_TypeHost_standard <- vcv_TypeHost  /(det(vcv_TypeHost  )^(1/nrow(vcv_TypeHost )))
```

```{r}

mal_dat_asex$obs <- c(seq(1, nrow(mal_dat_asex))) #each observation gets its own unique id

mal_dat_asex$Host_name <- as.factor(mal_dat_asex$Host)
mal_dat_asex$Parasite_name <- as.factor(mal_dat_asex$Plasmodium_species)
mal_dat_asex$Upper_Burst_Size <- as.integer(mal_dat_asex$Upper_Burst_Size)

```

The full model:

```{r message=FALSE, warning=FALSE, eval = FALSE, echo = TRUE}
model_null_ecophylo <- brm(
  Upper_Burst_Size ~ 1 +
    (1 | gr(Plasmodium_species, cov = vcv_Plasmodium_standard)) +
    (1 | gr(Host, cov = vcv_TypeHost_standard)) +
    (1 | Host_name) +
    (1 | Parasite_name) +
    (1 | obs),
  prior = c(
    prior(student_t(4, 0, 1), "sd",lb = 0, ub = 3)),
  data = mal_dat_asex,
  family = poisson("log"),
  warmup = 2000,
  iter = 4000,
  control = list(adapt_delta = 0.99,
                 max_treedepth = 11),
  data2 = list(
    vcv_Plasmodium_standard = vcv_Plasmodium_standard,
    vcv_TypeHost_standard = vcv_TypeHost_standard),
  save_pars = save_pars(all = TRUE)
)
```

```{r}
summary(fit_orig)
```

## Model validation

We then do a 'loo' model validation (leave-one-out). When we first calculate the loo for the original model, we see that 21% of the observation have low Pareto k diagnostic values. The bad diagnosis just indicates that we have to use the longer, more computationally-intense method where we fit the model for the 'bad' individual observations as the approximation (using the "Pareto distribution") just cannot approximate it.

```{r, echo = TRUE, eval = FALSE}
fit_orig <- add_criterion(model_null_ecophylo , "loo", reloo =
TRUE)
```

```{r}
print(fit_orig$criteria$loo)
```


Additionally, we check for the posterior predictive distribution
(pp_check). The black line is the observed maximum observed burst size
while the faint blue lines describe the random draws from the posterior
distribution. For both models, they seem to fit well.

```{r}
pp_check(model_null_ecophylo,type= "intervals",ndraws= 1000)
```

## Variance partitioning

To investigate if there is a phylogenetic signal, we propose two
methods: (1) comparing the loo of models only containing the host and
parasite phylogenies and ecologies, (2) and by calculating the variance attributed to
either host and parasite phylogeny over the total variance.

#### Loo

```{r, echo = TRUE, eval = FALSE}
model_parasites_only_ecophylo <- brm(
  Upper_Burst_Size ~ 1 +
    (1|gr(Plasmodium_species , cov = vcv_Plasmodium_standard))+
    (1|Parasite_name)+
    (1|obs),
  data = mal_dat_asex ,
  family = poisson('log'), 
  prior = c(
  prior(student_t(4, 0, 1), "sd",lb = 0, ub = 3)),
  warmup = 2000,
  iter = 10000,
  control = list(adapt_delta = 0.999),
  data2 = list(vcv_Plasmodium_standard = vcv_Plasmodium_standard,
               vcv_TypeHost_standard = vcv_TypeHost_standard),
  save_pars = save_pars(all = TRUE))
```

```{r}
summary(model_parasites_only_ecophylo)
```

```{r, echo = TRUE, eval = FALSE}
model_host_only_ecophylo <- 
  brm( Upper_Burst_Size ~ 1 +
  (1|gr(Host , cov = vcv_TypeHost_standard))+
  (1|Host_name)+
  (1|obs),
  data = mal_dat_asex, family = poisson('log'), 
  warmup = 2000, 
  iter =10000, 
  prior = c(prior(student_t(4, 0, 1), "sd",lb = 0, ub = 3)),
  control = list(adapt_delta = 0.999),
  data2 =list(vcv_Plasmodium_standard = vcv_Plasmodium_standard,
  vcv_TypeHost_standard = vcv_TypeHost_standard), 
  save_pars = save_pars(all = TRUE))
```

```{r, echo = TRUE, eval = FALSE}
fit_par_only<- add_criterion(model_parasites_only_ecophylo, "loo", reloo =
TRUE)
fit_host_only <- add_criterion(model_host_only_ecophylo,
"loo",reloo = TRUE)

```

```{r}
loo_compare(fit_orig,
            fit_par_only,
            fit_host_only,
            criterion = "loo")
```

The original model (fit_simp) containing both the host and parasite
phylogenies is the most similar to the model containing only the
parasite phylogeny (fit_par_only). The fit_host_only is significantly
different from the first two models.

#### Host and Parasite Effects

We borrow code from the {brms} forum, where we draw from the posterior
distribution to obtain the standard deviations of the random effects.
These are then squared to get the variances. The total variance is then
quantified by summing all the variances. Each random effects are then
divided by the total variance to get the proportion of the variance
explained.

```{r}
variance_partition <-fit_orig %>% 
  tidy_draws() %>% 
  dplyr::select(starts_with("sd_")) %>%
  transmute_all(.funs = list(sq = ~(. ^ 2))) %>% 
  mutate(total_var = rowSums(.)) %>%
  mutate_at(.vars = vars(-total_var), 
            .funs = list(pct = ~(. / total_var)))  %>% 
  map_df(.f = ~ median_hdci(., .width = 0.95), .id = "component") 
  
```

```{r}
ggplot(variance_partition[c(7,8,10,11),], aes(x = component, y = y))+
  geom_point(size =4) + 
  geom_segment(aes(x = component, xend=component, y=ymin, 
                                  yend=ymax), linewidth = 1.5)+
  scale_x_discrete(label=c("Host phylogeny",
                           "Host ecology",
                            "Parasite ecology",
                           "Parasite phylogeny"
                          ))+ 
  theme_classic()+
  xlab("Proportion of total variance")+
  ylab("Proportion of total variance")+
  theme(axis.text = element_text(size =14, color = 'black'),
        axis.title = element_text(size = 15, color = 'black'))
```

This plot shows the credible intervals for the proportion of variance
explained by either the host or parasite phylogeny. One can see that
parasite phylogeny is the most important driver

## Asexual development time
Using the same method above, we then add a covariate, the asexual duration.

```{r}
model_parasites_duration<- brm(
  Upper_Burst_Size ~ Duration +
    (1|gr(Plasmodium_species , cov = vcv_Plasmodium_standard))+
     (1|gr(Host , cov = vcv_TypeHost_standard))+
     (1|Host_name)+
    (1|Parasite_name)+
    (1|obs),
  data = mal_dat_asex ,
  family = poisson('log'), 
  prior = c(
  prior(student_t(4, 0, 1), "sd",lb = 0, ub = 3)),
  warmup = 2000,
  iter = 15000,
  control = list(adapt_delta = 0.999,
                 max_treedepth =15),
  data2 = list(vcv_Plasmodium_standard = vcv_Plasmodium_standard,
               vcv_TypeHost_standard = vcv_TypeHost_standard),
  save_pars = save_pars(all = TRUE))
```

```{r}
model_parasites_duration<- brm(
  Duration ~
    (1|gr(Plasmodium_species , cov = vcv_Plasmodium_standard))+
     (1|gr(Host , cov = vcv_TypeHost_standard))+
     (1|Host_name)+
    (1|Parasite_name),
  data = mal_dat_asex ,
  family = gaussian, 
  warmup = 2000,
  iter = 5000,
  control = list(adapt_delta = 0.90,
                 max_treedepth =15),
  data2 = list(vcv_Plasmodium_standard = vcv_Plasmodium_standard,
               vcv_TypeHost_standard = vcv_TypeHost_standard),
  save_pars = save_pars(all = TRUE))
```